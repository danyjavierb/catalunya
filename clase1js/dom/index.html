<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Document</title>
    <script defer src="./main.js"></script>
  </head>
  <body>
    <header>
      <h1 id="title">Tareas</h1>
    </header>
    <main class="mainContent">
      <!-- <table>
        <thead>
          <tr>
            <th>Id</th>
            <th>Nombre</th>
            <th>Edad</th>
            <th>Hobbies</th>
          </tr>
        </thead>
        <tbody id="cuerpoTabla"></tbody>
      </table> -->
      <input type="text" id="nombre" />
      <button id="crear">Crear Elemento</button>
    </main>
    <footer></footer>
  </body>
</html>
