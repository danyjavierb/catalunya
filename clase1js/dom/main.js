const estudiantes = [
  {
    id: 1,
    nombre: "Zee",
    edad: 20,
    hobbies: ["programar", "bailar"],
  },
  {
    id: 2,
    nombre: "Jassem",
    edad: 15,
    hobbies: ["programar", "jugar LoL"],
  },
  {
    id: 3,
    nombre: "Pierre",
    edad: 20,
    hobbies: ["enseñar", "programar"],
  },
  {
    id: 4,
    nombre: "Yeins",
    edad: 17,
    hobbies: ["cocinar"],
  },
];

// traer elemento por ID
const title = document.getElementById("title");

// traer elementos  por nombre de clase
const parrafos = document.getElementsByClassName("parrafo");

// traer elementos por tag name

const parrafosP = document.getElementsByTagName("p");

// usando selectores de css

const elemById = document.querySelector("#title");

const elemByClass = document.querySelector(".parrafo");

const elemByTagName = document.querySelector("p");

// especificidad
// pueden usar selectores de css para traer elementos del DOM

const ejem1 = document.querySelector("main p");

const ejem2 = document.querySelector("p.rojo");

const ejem3 = document.querySelector("p.parrafo");

console.log(ejem3);

const ejem4 = document.querySelector(".parrafo.rojo");

// todos los elementos que coincidan con el selector

const ejem5 = document.querySelectorAll("p.parrafo");
console.log(ejem5);

// modificando elementos

document.querySelector("#title").innerHTML = "Star Wars";

document.querySelector("#title").innerHTML =
  "<a href='google.com'>Ir a google</a>";

document.querySelector("#title a").href = "http://facebook.com";

// const elementoSeleccinado = document.querySelector(".parrafo.rojo");

// elementoSeleccinado.style.border = "solid 1px red";
// elementoSeleccinado.style.color = "green";

// Agregar elementos nuevos

// const nombresEstudiantes = estudiantes.map((est) => est.nombre);

// const estudianteNombreH1 = nombresEstudiantes.map((nombre) => {
//   const div = document.createElement("div");
//   div.innerHTML = `<h1> ${nombre} </h1>`;
//   return div;
// });

// const mainContent = document.querySelector(".mainContent");

// estudianteNombreH1.forEach((elem) => {
//   mainContent.appendChild(elem);
// });

// Pinten una tabla
// id   ||  nombre || edad    || hobbies
// 1    ||  Dany   || 30 ańos ||  . cantar . bailar

const table = document.createElement("table");
const estudianteRows = estudiantes.map((estudiante) => {
  const hobbieString = estudiante.hobbies.map((hobbie) => `<li>${hobbie}</li>`);
  const tr = `<tr> <td> ${estudiante.id} </td>
  <td> ${estudiante.nombre} </td>
  <td> ${estudiante.edad} años</td>
  <td> <ul>${hobbieString.join("")}</ul> </td>`;
  return tr;
});

const mainContent2 = document.querySelector(".mainContent");

table.innerHTML =
  "<tr> <td> ID </td> <td> NOMBRE </td> <td> EDAD </td> <td> HOBBIES </td></tr>" +
  estudianteRows.join("");

mainContent2.appendChild(table);

// Carlos

// const $cuerpoTabla = document.querySelector("#cuerpoTabla");
// // Recorrer todos los productos
// estudiantes.forEach((estudiantes) => {
//   // Crear un <tr>
//   const $tr = document.createElement("tr");
//   // Creamos el <td> de id y lo adjuntamos a tr
//   let $tdId = document.createElement("td");
//   $tdId.textContent = estudiantes.id; // el textContent del td es el id
//   $tr.appendChild($tdId);
//   // Creamos el <td> de nombre y lo adjuntamos a tr
//   let $tdNombre = document.createElement("td");

//   $tdNombre.textContent = estudiantes.nombre; // el textContent del td es el nombre
//   $tr.appendChild($tdNombre);
//   // El td de edad
//   let $tdPrecio = document.createElement("td");
//   $tdPrecio.textContent = estudiantes.edad;
//   $tr.appendChild($tdPrecio);
//   // El td del hobbies
//   let $tdCodigo = document.createElement("td");
//   $tdCodigo.textContent = estudiantes.hobbies;
//   $tr.appendChild($tdCodigo);

//   $cuerpoTabla.appendChild($tr);
// });

const button = document.querySelector("#crear");

const crearElementoHanlder = (ev) => {
  const div = document.createElement("div");
  const nombreInput = document.querySelector("#nombre").value;
  div.innerHTML = `<h1> HOLA ${nombreInput} </h1>`;
  document.querySelector(".mainContent").appendChild(div);
};

button.addEventListener("click", crearElementoHanlder);
