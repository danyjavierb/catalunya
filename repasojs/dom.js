// interactuar con el arbol de elementos del sitio

// Seleccionar elementos del dom

//seleccion del primer elemento que coincida con el selector
let divSaludo = document.querySelector(".saludo");
console.log(divSaludo);

let todosLosDivsSaludos = document.querySelectorAll(".saludo");
console.log(todosLosDivsSaludos);

let elementoPorId = document.querySelector("#lista");
console.log(elementoPorId);

// atributos de un elemento

let inputNumero = document.querySelector("#numero1");

console.log(parseInt(inputNumero.value));

inputNumero.style = "border: red solid 1px";
inputNumero.value = 123123;

inputNumero.classList.add("colorFont");

todosLosDivsSaludos.forEach((div) => {
  //div.classList.add("colorFont");
});

inputNumero.classList.remove("colorFont");

// crearlos

let nuevoElementoDiv = document.createElement("div");
//agregar contenido a un elemento

nuevoElementoDiv.innerHTML = `<h1>H1 interno</h1>`;

todosLosDivsSaludos[1].appendChild(nuevoElementoDiv);

const programadores = [
  {
    id: 1,
    nombre: "Dany Bautista",
    edad: 20,
    lenguajes: ["java", "js", "scala"],
    imprimirPres: () => {
      return "dany es un ingeniero de sofware.....";
    },
  },
  {
    id: 2,
    nombre: "Camila Sarmiento",
    edad: 30,
    lenguajes: ["js", "ts", "rescript"],
    imprimirPres: () => {
      return "Camila Sarmiento es programadora fronent.....";
    },
  },
  {
    id: 3,
    nombre: "Jorge Castellanos",
    edad: 30,
    lenguajes: ["Python", "Nodejs", "Ruby", "java"],
    imprimirPres: () => {
      return "Jorge es un programador backend.....";
    },
  },
];

const traerNombresProgramadores = (arr) => {
  let salida = [];
  for (i = 0; i < arr.length; i++) {
    salida.push(arr[i]["nombre"]);
  }
  return salida;
};

let nombres = traerNombresProgramadores(programadores);

let listaHtml = document.querySelector("#lista");

programadores.forEach((prog) => {
  let liNombre = document.createElement("li");
  liNombre.classList.add("colorFont");
  liNombre.innerHTML = `
  <a href="#" > ${prog.nombre} </a>
  </br>
  <span>${prog.imprimirPres()} </span>
  `;
  listaHtml.appendChild(liNombre);
});

//elementos les podemos agregar escuchadores de eventos

let botonCalcular = document.querySelector("#calcular");

botonCalcular.addEventListener("click", () => {
  let value1 = parseInt(document.querySelector("#numero1").value);
  let value2 = parseInt(document.querySelector("#numero2").value);
  let resultadoCalculo = value1 + value2;
  console.log("calculando");

  document.querySelector("#resultado").value = resultadoCalculo;
});

//hacer una calculadora en html js
