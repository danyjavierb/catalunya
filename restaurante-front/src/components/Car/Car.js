import React, { useState } from "react";
import { Dash } from "react-bootstrap-icons";
import { Plus } from "react-bootstrap-icons";
import { Nav, Button } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import { addUnit, emptyCart, removeUnit } from "../../state/cart.duck";
import { enviarPedidoAction } from "../../state/pedidos.duck";

const Car = (props) => {
  const message = "No hay items en el carrito de compras";

  const [formaPago, setFormaPago] = useState("DATAFONO");

  const carItems = useSelector((state) => state.cart);
  const dispatch = useDispatch();

  const handleAddUnit = (dish) => {
    dispatch(addUnit(dish));
  };

  const handleRemoveUnit = (dish) => {
    dispatch(removeUnit(dish));
  };

  const handleOnSubmitOrder = () => {
    dispatch(enviarPedidoAction(carItems, formaPago)).then((nuevoPedido) => {
      alert(`pedido creado con id ${nuevoPedido.id}`);
      dispatch(emptyCart());
    });
  };

  const getTotal = () => {
    return carItems.reduce((acc, item) => {
      return acc + item.precio * item.cantidad;
    }, 0);
  };

  if (carItems.length == 0) {
    return <h5>{message}</h5>;
  }

  return (
    <Nav
      className="col-md-2 d-none d-md-block bg-light sidebar"
      activeKey="/home"
      onSelect={(selectedKey) => alert(`selected ${selectedKey}`)}
    >
      <div className="sidebar-sticky" />
      {carItems.map((item) => {
        return (
          item.cantidad > 0 && (
            <Nav.Item key={`${item.nombre}-${item.cantidad}`}>
              <h6>{item.nombre}</h6>
              <span>Cantidad: {item.cantidad}</span>
              <Dash
                data-testid={`remove_${item.id}`}
                onClick={() => handleRemoveUnit(item)}
              />
              <Plus
                data-testid={`add_${item.id}`}
                onClick={() => handleAddUnit(item)}
              />
              <br />
              <br />
            </Nav.Item>
          )
        );
      })}
      <div>Total: {carItems.length && getTotal()}</div>
      <div>
        Forma de pago:
        <select
          onChange={(ev) => {
            setFormaPago(ev.target.value);
          }}
        >
          <option value="DATAFONO">DATAFONO</option>
          <option value="EFECTIVO">EFECTIVO</option>
        </select>
      </div>
      <Button onClick={() => handleOnSubmitOrder()}>Hacer pedido</Button>
    </Nav>
  );
};

export default Car;
