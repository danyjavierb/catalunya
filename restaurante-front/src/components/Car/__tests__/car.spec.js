import { render, fireEvent, waitFor, getByText } from "@testing-library/react";
import { screen } from "@testing-library/dom";
import "@testing-library/jest-dom/extend-expect";
import Car from "../Car";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe("tests carrito", () => {
  const ADD_UNIT_TYPE = "ADD_UNIT";
  const REMOVE_UNIT_TYPE = "REMOVE_UNIT";

  const store = mockStore({
    cart: [
      {
        id: 1,
        nombre: "hamburguesa sencilla",
        precio: 10000,
        cantidad: 2,
      },
    ],
  });

  it("Carrito debe mostrar precio total", () => {
    const { container } = render(
      <Provider store={store}>
        <Car />
      </Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
    expect(screen.getByText("Total: 20000")).toBeInTheDocument();
  });

  it("al presionar el boton + se debe despachar ADD_UNIT_TYPE", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Car />
      </Provider>
    );

    const expectedAction = [
      {
        type: ADD_UNIT_TYPE,
        payload: {
          id: 1,
          nombre: "hamburguesa sencilla",
          precio: 10000,
          cantidad: 2,
        },
      },
    ];

    const button = getByTestId("add_1");
    fireEvent.click(button);
    expect(store.getActions()).toEqual(expectedAction);
  });
});
