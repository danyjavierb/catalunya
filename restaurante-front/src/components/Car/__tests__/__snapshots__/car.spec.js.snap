// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests carrito Carrito debe mostrar precio total 1`] = `
<div
  class="col-md-2 d-none d-md-block bg-light sidebar nav"
>
  <div
    class="sidebar-sticky"
  />
  <div
    class="nav-item"
  >
    <h6>
      hamburguesa sencilla
    </h6>
    <span>
      Cantidad: 
      2
    </span>
    <svg
      data-testid="remove_1"
      fill="currentColor"
      height="1em"
      viewBox="0 0 16 16"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M4 8a.5.5 0 0 1 .5-.5h7a.5.5 0 0 1 0 1h-7A.5.5 0 0 1 4 8z"
      />
    </svg>
    <svg
      data-testid="add_1"
      fill="currentColor"
      height="1em"
      viewBox="0 0 16 16"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 4a.5.5 0 0 1 .5.5v3h3a.5.5 0 0 1 0 1h-3v3a.5.5 0 0 1-1 0v-3h-3a.5.5 0 0 1 0-1h3v-3A.5.5 0 0 1 8 4z"
      />
    </svg>
    <br />
    <br />
  </div>
  <div>
    Total: 
    20000
  </div>
  <div>
    Forma de pago:
    <select>
      <option
        value="DATAFONO"
      >
        DATAFONO
      </option>
      <option
        value="EFECTIVO"
      >
        EFECTIVO
      </option>
    </select>
  </div>
  <button
    class="btn btn-primary"
    type="button"
  >
    Hacer pedido
  </button>
</div>
`;
